File: journey_to_the_moon.go
package main

import "fmt"

type QuickFind struct {
	arr []int
}

func (q *QuickFind) Union(i, j int) {
	p1 := q.arr[i]
	p2 := q.arr[j]

	for k := 0; k < len(q.arr); k++ {
		if q.arr[k] == p2 {
			q.arr[k] = p1
		}
	}
}

func (q *QuickFind) AreConnected(i, j int) bool {
	return q.arr[i] == q.arr[j]
}

func (q *QuickFind) Elements() []int {
	return q.arr
}

func newQuickFind(n int) *QuickFind {
	arr := make([]int, n)

	for i := 0; i < len(arr); i++ {
		arr[i] = i
	}

	a := QuickFind{arr: arr}
	return &a
}

func main() {
	var N, I int
	fmt.Scanf("%d %d", &N, &I)

	finder := newQuickFind(N)

	for i := 0; i < N; i++ {
		var p1, p2 int
		fmt.Scanf("%d %d", &p1, &p2)
		finder.Union(p1, p2)
	}

	counter := 0
	elements := finder.Elements()

	fmt.Print(elements)

	for i := 0; i < N; i++ {
		for j := i + 1; j < N; j++ {
			if elements[i] != elements[j] {
				counter++
			}
		}
	}
	fmt.Println(counter)
}

File: journey_to_the_moon.py
from collections import Counter, defaultdict

class QuickFind:
    def __init__(self, size):
        self._arr = [i for i in range(size)]

    def are_connected(self, p1, p2):
        return self._arr[p1] == self._arr[p2]

    def union(self, p1, p2):
        p1_id = self._arr[p1]
        p2_id = self._arr[p2]

        for index, element in enumerate(self._arr):
            if element == p1_id:
                self._arr[index] = p2_id

    def elements_groups(self):
        return self._arr

class QuickUnion:
    def __init__(self, size):
        self._arr = [i for i in range(size)]
        self.size = size

    def root(self, element):
        while(element != self._arr[element]):
            element = self._arr[element]
        return element

    def union(self, p1, p2):
        p1_root = self.root(p1)
        p2_root = self.root(p2)

        self._arr[p2_root] = p1_root

    def elements_groups(self):
        return [self.root(e) for e in self._arr]

    def are_connected(self, p1, p2):
        return self.root(p1) == self.root(p2)

arr = [
	[0, 2],
	[1, 8],
	[1, 4],
	[2, 8],
	[2, 6],
	[3, 5],
	[6, 9],
]


a = QuickFind(10)

for i, j in arr:
    a.union(i, j)

elements = a.elements_groups()
print(sum(1 for i in range(10) for j in range(i + 1, 10) if elements[i] != elements[j]))
